new-log-viewer/src/components/Editor/MonacoInstance/actions.ts

```diff
@@ -0,0 +1,168 @@
+import * as monaco from "monaco-editor/esm/vs/editor/editor.api.js";
+
+import {Nullable} from "../../../typings/common";
+import {ActionType} from "../../../utils/actions";
+import {clamp} from "../../../utils/math";
+import type {
+    CursorExplicitPosChangeCallback,
+    CustomActionCallback,
+} from "./typings";
+
+
+const MIN_ZOOM_LEVEL = 1;
+const MAX_ZOOM_LEVEL = 10;
+const MOBILE_ZOOM_LEVEL_INCREMENT = 10;
+const MOBILE_ZOOM_LEVEL_DECREMENT = 1;
+const POSITION_CHANGE_DEBOUNCE_TIMEOUT_MILLIS = 50;
+
+/**
+ * Sets up an action that is triggered when the cursor position changes in a Monaco code editor.
+ *
+ * @param editor
+ * @param onCursorExplicitPosChange
+ */
+const setupCursorExplicitPosChangeAction = (
+    editor: monaco.editor.IStandaloneCodeEditor,
+    onCursorExplicitPosChange: CursorExplicitPosChangeCallback
+) => {
+    let posChangeDebounceTimeout: Nullable<ReturnType<typeof setTimeout>> = null;
+
+    editor.onDidChangeCursorPosition((ev: monaco.editor.ICursorPositionChangedEvent) => {
+        // only trigger if there was an explicit change that was made by keyboard or mouse
+        if (monaco.editor.CursorChangeReason.Explicit !== ev.reason) {
+            return;
+        }
+        if (null !== posChangeDebounceTimeout) {
+            clearTimeout(posChangeDebounceTimeout);
+        }
+        posChangeDebounceTimeout = setTimeout(() => {
+            onCursorExplicitPosChange(ev);
+            posChangeDebounceTimeout = null;
+        }, POSITION_CHANGE_DEBOUNCE_TIMEOUT_MILLIS);
+    });
+};
+
+/**
+ * Sets up editor focus on `backtick` key down.
+ *
+ * @param editor
+ */
+const setupFocusOnBacktickDown = (editor: monaco.editor.IStandaloneCodeEditor) => {
+    const handleKeyDown = (e: KeyboardEvent) => {
+        if ("`" === e.key) {
+            e.stopPropagation();
+            e.preventDefault();
+            editor.focus();
+        }
+    };
+
+    window.addEventListener("keypress", handleKeyDown);
+    editor.onDidDispose(() => {
+        window.removeEventListener("keypress", handleKeyDown);
+    });
+};
+
+/**
+ * Calculates the distance between two touch points.
+ *
+ * @param touch1
+ * @param touch2
+ * @return The Euclidean distance between the touch points.
+ */
+const getTouchDistance = (touch1: Touch, touch2: Touch): number => Math.sqrt(
+    ((touch2.pageX - touch1.pageX) ** 2) + ((touch2.pageY - touch1.pageY) ** 2)
+);
+
+/**
+ * Sets up mobile zoom functionality by calculating distance differences between two touch points.
+ *
+ * @param editor
+ * @param editorContainer
+ */
+const setupMobileZoom = (
+    editor: monaco.editor.IStandaloneCodeEditor,
+    editorContainer: HTMLElement
+) => {
+    const editorDomNode = editor.getDomNode();
+    if (null === editorDomNode) {
+        console.error("Unexpected null returned by editor.getDomNode()");
+
+        return;
+    }
+
```

Comment:
```suggestion

    // NOTE: We explicitly set `passive=false` for the listeners below since it defaults to `true`
    // in Safari for touch events (whereas it defaults to `false` otherwise).

```